src/tools/kotlin-schema-generator.ts

```diff
@@ -8,83 +8,102 @@
-import {AddFieldOptions, SchemaDescriptorBase} from './schema2base.js';
 import {KTExtracter} from './kotlin-refinement-generator.js';
-import {escapeIdentifier, getTypeInfo} from './kotlin-codegen-shared.js';
+import {assert} from '../platform/assert-web.js';
 
 const ktUtils = new KotlinGenerationUtils();
 
 /**
- * Metadata about a field in a schema.
+ * Generates a Kotlin schema instance.
  */
-export type KotlinSchemaField = AddFieldOptions & {
-  type: string,
-  decodeFn: string,
-  defaultVal: string,
-  schemaType: string,
-  escaped: string,
-  nullableType: string
-};
-
-/**
- * Composes and holds a list of KotlinSchemaField for a SchemaNode.
- */
-export class KotlinSchemaDescriptor extends SchemaDescriptorBase {
-
-  readonly fields: KotlinSchemaField[] = [];
-
-  constructor(node: SchemaNode, private forWasm: boolean) {
-    super(node);
-    this.process();
-  }
-
-  addField(opts: AddFieldOptions) {
-    if (opts.typeName === 'Reference' && this.forWasm) return;
-
-    const typeInfo = getTypeInfo({name: opts.typeName, ...opts});
-    const type = typeInfo.type;
-
-    this.fields.push({
-      ...opts,
-      ...typeInfo,
-      escaped: escapeIdentifier(opts.field),
-      nullableType: type.endsWith('?') ? type : `${type}?`
-    });
-  }
-}
-
-/**
- * Generates a schema instance for a given KotlinSchemaDescriptor
- */
-export function generateSchema(descriptor: KotlinSchemaDescriptor): string {
-  const schema = descriptor.node.schema;
+export async function generateSchema(schema: Schema): Promise<string> {
   if (schema.equals(Schema.EMPTY)) return `Schema.EMPTY`;
 
   const schemaNames = schema.names.map(n => `SchemaName("${n}")`);
   const {refinement, query} = generatePredicates(schema);
 
-  function generateFieldMap(isCollection: boolean): string {
-    const fieldPairs = descriptor.fields
-        .filter(f => !!f.isCollection === isCollection)
-        .map(({field, schemaType}) => `"${field}" to ${schemaType}`);
-    return leftPad(ktUtils.mapOf(fieldPairs, 30), 8, true);
-  }
+  const singletons: string[] = [];
+  const collections: string[] = [];
+
+  await visitSchemaFields(schema, ({isCollection, field, schemaType}) =>
+      (isCollection ? collections : singletons).push(`"${field}" to ${schemaType}`));
 
   return `\
 Schema(
     setOf(${ktUtils.joinWithIndents(schemaNames, {startIndent: 8})}),
     SchemaFields(
-        singletons = ${generateFieldMap(false)},
-        collections = ${generateFieldMap(true)}
+        singletons = ${leftPad(ktUtils.mapOf(singletons, 30), 8, true)},
+        collections = ${leftPad(ktUtils.mapOf(collections, 30), 8, true)}
     ),
-    ${quote(descriptor.node.hash)},
+    ${quote(await schema.hash())},
     refinement = ${refinement},
     query = ${query}
 )`;
 }
 
+interface SchemaField {
+  field: string;
+  isCollection: boolean;
+  schemaType: string;
+}
+
+async function visitSchemaFields(schema: Schema, visitor: (field: SchemaField) => void) {
```

Comment:
Is it possible that some of this logic could be shared to the `EntityDescriptorBase.process` logic?  Maybe for a future PR...

Reply:
This is where we started, all this stuff started in Entity(Schema)DescriptorBase, but that was wrong, as we were handling both the schema generation and entity generation in one place and that's what took my multiple hours last night to untangle. These are different things and should be separate. The fact that in both cases we analyze the schema data structure does not mean that we should reuse code, if reusing code would mean obfuscating the logic and/or making it very hard to reuse logic, which is what we had.

If anything, IMO the `EntityDescriptorBase.process` should be cleaned up, to resemble more what I've done here - which was produced by carefully refactoring `EntityDescriptorBase.process` to reveal the inner structure of schema fields descriptors.